RFC: Monitoring across the stack

sarayourfriend opened the pull request:
Additionally some preliminary sketched out code as a starting point

Respond by: 2022-03-25

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

sarayourfriend commented on justfile:
```diff
@@ -36,3 +36,45 @@ _py-install:
+@monitor-logs service="":
+    just _monitor-dc "logs {{ service }}"
+
+@remonitor:
+    just _monitor-dc "down -v"
+    just monitor "--force-recreate --build"
```
```suggestion
@remonitor environment="local":
    just _monitor-dc "down -v"
    just monitor {{ environment }} "--force-recreate --build"
```

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+### Local infrastructure recommendations
+
+One of the goals I'd like to achieve with the local infrastructure is to make it as easy as running `just monitor` and it spins up everything with tolerances for services that aren't currently running locally. For example, if someone is working primarily on developing frontend monitors, it would do no good for Prometheus to be producing errors about not be able to reach a local API, Airflow, imag ...
```
Totally agree. Do you think it would be a reasonable requirement to have folks run `just monitor` from the Wordpress/openverse repository?

sarayourfriend replied:
If they're working on monitoring yes.

obulat replied:
Love this!

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+Pros for Prometheus:
+- It is possible and potentially even trivial to provision alerts based on configuration files
+Cons for Prometheus:
```
Small formatting fix
```suggestion

Cons for Prometheus:
```

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+Pros for Grafana:
+- The UI for building them is nice
+Cons for Grafana:
```
Same as above
```suggestion

Cons for Grafana:
```

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+Cons for Grafana:
+- There's no way to export or provision them so the only way to add rules is through the UI (there's no JSON version of the form either). Moving rules across environments would be hard... to be fair, developing rules with anything except production data is also very difficult (lots of guessing) so maybe this isn't a problem.
+
+I'd rather use Grafana provided we back up the database of alert configurations. If we do that then not having them in the code is hopefully less of a disaster.
```
Same, I'd much rather try to fiddle with this in the UI than do an iterative process _only for production_ where we have to edit config files.

stacimc replied:
Same for me as well.

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+* Rely on Grafana for visbility and alarming upstream from Prometheus.
+* Create separate alarm streams per service in Mailman and aggregate all alarms into a Slack channel for extra visibility and redundancy.
+* Use `django-prometheus` library to add the Prometheus `/metrics` endpoint to the API
+* Expose Grafana dashboards publicly _a la_ [Wikimedia](https://grafana.wikimedia.org/?orgId=1) and other FOSS projects
```
WOAH

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+Graphite is also older than Prometheus and while it works really well, there's a lot less current information about how to deploy and configure it. There's also almost no support for languages outside of the core langauges supported by the project including poor JavaScript support (which would make monitoring our Nuxt service more difficult).
+
+### CloudWatch
```
I agree with all these points and would love to NOT use CloudWatch where possible :)

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+### The ELK stack
+
+The final alternative is the [ELK stack](https://www.elastic.co/what-is/elk-stack). It's not exactly a 1:1 alternative to Prometheus and Grafana, though it could fulfill [some](https://www.elastic.co/guide/en/kibana/current/xpack-ml-anomalies.html) of the [same](https://www.elastic.co/kibana/) use cases. This is a strong alternative that I hope reviewers of this RFC spend time considering. Read t ...
```
I cannot say I have a ton of experience with the ELK stack, but what I remember wasn't too enjoyable. Kibana was difficult to grok and navigate. I remember our ops team at a previous job having lots of trouble setting it all up. That said, all of this is tertiary and I don't have any specific, data-backed resistance to it.

sarayourfriend replied:
I don't think Kibana's queries are all that much more cryptic than Prometheus, FWIW. Prometheus might be even more so, it's extremely syntax heavy whereas Kibana has a tendency to be pretty verbose.

AetherUnbound reviewed (commented):
This is an incredible outline, thank you for all the work and consideration you put into this Sara! I have a couple comments/questions/suggestions.

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+## Purpose
+
+This RFC proposes we introduce [Prometheus](https://prometheus.io) and [Grafana](https://grafana.com/) to monitor our infrastructure. It also proposes that we use our existing Postgres instance for storing the gathered metrics.
```
I'm pretty sure we only have a single RDS instance based on what I'm seeing in our terraform modules.

stacimc replied:
Aside from the staging db, I see prod and upstream (Catalog). I'm not sure if there's an obvious choice between the two for storing metrics, though. Would there be any benefit to separating it from the API db?

zackkrida replied:
That's correct, there are three, one for the staging catalog, one for the prod catalog, and one for the prod api.

I don't think cost should be a deciding factor here; there are other places we can make cost optimizations if we need to.

sarayourfriend replied:
Cool, let's make a new RDS for it then.

AetherUnbound replied:
> we only have a single RDS instance based on what I'm seeing in our terraform modules.

Ah, that just speaks to the reality that our terraform modules don't fully encapsulate all of our AWS resource, unfortunately :sweat: +1 to a new RDS instance for this.

zackkrida replied:
Can we update the RFC to reflect this discussion and resolve this thread, @sarayourfriend ?

sarayourfriend replied:
I've already updated it to recommend using RDS. Was there something else that needed to be updated @zackkrida?

krysal replied:
@sarayourfriend In line [34][lineref] it says:

> ... It also proposes that we use our existing Postgres instance for storing the gathered metrics.

+1 to a new RDS instance for this!


[lineref]: https://github.com/WordPress/openverse/blob/021533e5c528697e0d0bdfb8bf9c0c0942f8d8b3/rfcs/20220307-monitoring_across_the_stack.md?plain=1#L34

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+## Self-hosted vs. hosted
+
+I will defer this decision to @rbadillap because he has way more knowledge about what will be affordable for us both from a raw cost perspective and from a time-spent-maintaining perspective. My gut says that in the interest of getting monitoring sooner than later we could pay for hosted Prometheus and Grafana to start with and eventually migrate to self-hosted. AWS has hosted offerrings for both ...
```
That's a good point.

I need to revise this section. Grafana by default uses SQLite (makes provisioning easy). I wonder if we can survive with just that for now.

I'd hate to spin up a second RDS instance just for monitoring, just from a cost perspective. @rbadillap do you have any thoughts here? Can we get away with using SQLite in production for Grafana or is that going to bite us eventually?

AetherUnbound replied:
My gut says that we should spin up a new instance and that it wouldn't cost _too_ much (if Grafana can survive on sqlite it can certainly survive on a small postgres instance).

sarayourfriend replied:
Yeah if we don't care about cost as suggested by @zackkrida then just use a new instance. It will be more resilient into the future anyway.

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+[Here is the list of Grafana "contact points" for alerting.](https://grafana.com/docs/grafana/latest/alerting/unified-alerting/contact-points/) A contact point is the notification channels that Grafana can send alerts to.
+
+The most relevant for our usage are probably Slack, Webhook, and Email. Optionally we could incorporate PagerDuty depending on the team's tolerance and how we decide to organize expectations around responding to alerts.
```
My hang up with PagerDuty is that it's not free and requires additional accounts, therefore additional security practices, etc.

Feature wise I think it's also overkill for our team's requirements.

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+Note: Because I had to spend a good deal of time sketching out how this stuff works, I wrote a good deal of preliminary code. It's probably not perfect so it's really only a starting point, but it at least gives a vision for how we could do this.
+
+Additionally, there are several 
```
Not sure what I was writing there. Removed.

stacimc commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,361 @@
+
+My preference is for us to primarily rely on email and in particular to use email lists to allow individuals to subscribe/unsubscribe at will from alarm streams and provide an easy historical public record of alarms. We could use something like [GNU Mailman](https://list.org/) to [host this ourselves](https://github.com/maxking/docker-mailman) using FARGATE.
+
+Email also makes it easy to compartmentalize alarms per category, it makes it easy for individuals to subscribe to specific streams of alarms by using different email prefixes per category. For example, each service could have its own mailing list.
```
💯 I'm sold! I think my preference would be email & Slack. +1 to all of this.

stacimc commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,361 @@
+Runbooks are the expected steps for addressing a particular alarm. As I mentioned elsewhere, alarms should be actionable. If they're not actionable then they're just noise and should be demoted to a warning until they're more accurate (less false positives). The _actions_ you take for any given alarm are directed by an alarm's accomanying runbook.
+
+All alarms should have a relevant runbook linked in the alarm body. Ideally these would just be flat Markdown files but it'd be best if they could be hosted in the Openverse handbook on WordPress.org as well for redundancy and accessibility.
+
```
+1 I'm glad you called this out in particular. This really exposes which alarms are meaningful and actionable. I'd add that it's also important for making it easier to distribute the work of _addressing_ alarms 💯

zackkrida replied:
This section is fantastic.

stacimc commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,361 @@
+
+Do we want to hide the metrics endpoint behind authentication? I don't really see why we would other than to stop spamming (but we could throttle those endpoints to match the prometheus scrape rate for example and it would prevent anyone from doing too much damage).
+
+I like the idea of having fully open Prometheus and Grafana dashboards that don't require authentication to _explore_. Obviously we'd want to hide administration inside of authentication but each Prometheus and Grafana handle that on their own.
```
+1, I think this would be very cool.

obulat replied:
I would like to have open dashboards.

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+We can also integrate Slack for good measure and have all alarms sent to the relevant channels via Slack incoming Webhook. It's very easy to configure the Slack integration.
+
+The general purpose Webhook integration is useful for a future where we migrate away from Slack to Matrix (should that ever happen).
```
I guess I could say "when that eventuality occurs" to be less cynical :sweat_smile:

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+* Cloudflare exporter: https://gitlab.com/gitlab-org/cloudflare_exporter
+* AWS CloudWatch exporter: https://github.com/prometheus/cloudwatch_exporter
+    * This one is interesting. I think we'd define system metrics to gather in CloudWatch and then export them into Prometheus. Hopefully this wouldn't incur additional costs to what we're already doing with CloudWatch.
+* Alternative to the CloudWatch exported is to implement the Node exporter: https://github.com/prometheus/node_exporter
```
Added, thanks!

sarayourfriend commented:
@zackkrida I've removed Ronny from the requested reviewers and replaced his infrastructure sign-off role with you, hope that's okay.

I still need two people from the team to approve this if it looks good @WordPress/openverse. Thanks!

AetherUnbound commented on monitoring/docker-compose.yml:
```diff
@@ -0,0 +1,23 @@
+version: "3"
```
Dang, that's unfortunate :cold_sweat:

sarayourfriend replied:
It is. I think for this, unfortunately, we will need to use AWS's (or some other) private container image registry.

zackkrida replied:
Yeah, I think AWS' ECR would be appropriate for this, unfortunately.

AetherUnbound commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+## Self-hosted vs. hosted
+
+We will seslf-host and use a new Postgres RDS instance for Grafana.
```
```suggestion
We will self-host and use a new Postgres RDS instance for Grafana.
```

AetherUnbound reviewed (approved):
Some very small notes, but I'm comfortable approving this! Looking forward to having some stats to check out :female_detective:

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+## Reviewers
+
+- [ ] @zackkrida (Infrastructure sign off)
+- [ ] Approver TBD (Catalog/API person)
```
```suggestion
- [x] @AetherUnbound
```

zackkrida commented on monitoring/docker-compose.yml:
```diff
@@ -0,0 +1,23 @@
+  grafana:
+    image: grafana/grafana:8.4.4
+    volumes:
+      - ./configuration/grafana:/var/lib/grafana:z
```
I think AWS' managed Postgres would be nice so the backup story is the same as our other databases, and if we use any db specific tooling or monitoring in the future we'd have consistency.

sarayourfriend replied:
Yup! This comment is out of date now, I changed the proposal to just use Postgres.

zackkrida commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+* Storage
+* Processes/threads
+
+These would be gathered from Airflow, API, Thumbnail proxy, Postgres and Nuxt. These metrics need to be retrieved from AWS. I _think_ that [this page describes how to do that](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-PrometheusEC2.html) but AWS's language around this stuff is so cryptic to me I can't say with confidence it's the right documentation page for  ...
```
I think this would work, yes. If we followed this approach, ee'd want to modify our existing user data scripts (the `init.tpl` files for each terraform service that define scripts to run when each instance is instantiated) to install the CloudWatch Agent with `sudo yum install amazon-cloudwatch-agent`. 

We'd then also need to attach a new policy (AWS-speak for some configuration that gives services access to each other) to each instance, a `CloudWatchAgentServerPolicy`. 

Alternatively, I'm seeing a lot of folks using a solution decoupled from AWS: Prometheus's [node_exporter](https://prometheus.io/docs/guides/node-exporter/). This is a static binary that works on any Linux server that exposes metrics over port 9100. I like what I'm reading about it.

Last note: our servers are _already using_ [Filebeat](https://www.elastic.co/beats/filebeat), which at one point was used to post logs to [graylog](https://www.graylog.org/). That's still installed on every instance and apparently _can_ work with Prometheus, although google searches on that really aren't clear.

zackkrida replied:
Ah, forgive me, you mention the node_exporter later in the RFC.

zackkrida commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+The first one also seems complicated but I believe it's also easier. Remember that Prometheus is the source of truth for the configuration, so we wouldn't have to spread out the knowledge about where Prometheus lives in all our services. As long as our services are configured to serve Prometheus the appropriate data on consistent HTTP endpoints, then any Prometheus instance anywhere can consume t ...
+
+I'm not sure where the best place to put the monitoring docker-compose is. I think I'd like to propose putting it in WordPress/openverse under a `monitoring` directory that would include the `docker-compose.yml` as well as documentation about how to configure monitors, our own best practices, etc. It doesn't make sense to me for us to just pick one of the three application repositories to put it  ...
```
Support this as well and hope that monorepo approval from the team, once that RFC lands, would make this even easier for folks to understand.

zackkrida commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+I've added some new `just` recipes in the example code in this PR.
+
+Grafana configurations are stored as JSON so I think they'd be linted by prettier. They should be copy/paste exports anyway so we don't need to lint them for correctness or even style technically.
+
+Prometheus configurations can be linted by `promtool` using this pre-commit repo: https://github.com/fortman/pre-commit-prometheus.
+
```
Really appreciate these linting considerations!

zackkrida commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+I'd rather use Grafana provided we back up the database of alert configurations. If we do that then not having them in the code is hopefully less of a disaster.
+
+It is possible to export _all_ of them via the rules API: `http://grafana/api/ruler/grafana/api/v1/rules`. There's just no way to upload that JSON blob directly. We could scrape the endpoint and back it up into the code base for good measure and to further distribute the configuration.
```
```
We could scrape the endpoint and back it up into the code base for good measure and to further distribute the configuration.
```

Love it!

zackkrida reviewed (approved):
Left some comments, I think this is really great. I do want to experiment with you at some point and check out our _existing but not functioning_ monitoring stack and see what's happening there and how it might overlap or be replaced by this.

Here's also a little tutorial I read while reviewing this that was insightful:

https://medium.com/investing-in-tech/automatic-monitoring-for-all-new-aws-instances-using-prometheus-service-discovery-97d37a5b2ea2

zackkrida commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+## Reviewers
+
+- [ ] @zackkrida (Infrastructure sign off)
```
```suggestion
- [x] @zackkrida (Infrastructure sign off)
```

obulat commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,353 @@
+
+Prometheus is a time series database with an advanced query language and the ability to configure aggregate data transformations (for example, based on a given metric you can also tell Prometheus to calculate a running mean of the metric from over the last 5 minutes, or any other span of time). It includes anomaly detection as well as integration with many tools you can use for alerting like Slac ...
+
+Grafana is a data visualization tool that supports building complex dashboards. It has the ability to integrate with nearly any source of data you could imagine, including our existing catalog and API databases. In the future we could use it to monitor cataloging progress over time for example. It's a tool good for many things in addition to visualizing metrics.
```
Me too! This would be a really interesting project.

obulat commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+These would be gathered from Airflow, API, Thumbnail proxy, Postgres and Nuxt. These metrics need to be retrieved from AWS. I _think_ that [this page describes how to do that](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch-Agent-PrometheusEC2.html) but AWS's language around this stuff is so cryptic to me I can't say with confidence it's the right documentation page for  ...
+
+It would also be nice if we could integrate Cloudflare data into Grafana so that our visualizations are all in once place. Cache hits and other relevant information from Cloudflare would be excellent to monitor within our greater monitoring stack.
```
```suggestion
It would also be nice if we could integrate Cloudflare data into Grafana so that our visualizations are all in one place. Cache hits and other relevant information from Cloudflare would be excellent to monitor within our greater monitoring stack.
```

obulat commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+Prometheus works by scraping HTTP endpoints on services that provide a formatted set of metrics data. Essentially services don't need to worry about whether Prometheus is running or concern themselves with how long to wait between sending batched event lists. Each service just writes to a file (or some other storage) that is served to Prometheus over HTTP whenever it requests it.
+
+That has some really nice properties in that each service doesn't need to worry about what Prometheus is doing. We don't need to worry about it's tolerances for getting new events, rate limits, handling bad requests or timeouts if Prometheus is temporarily unavailable, anything like that. We (or rather the client librariy) just write and serve the metric data in the appropriate format and Prometh ...
```
```suggestion
That has some really nice properties in that each service doesn't need to worry about what Prometheus is doing. We don't need to worry about it's tolerances for getting new events, rate limits, handling bad requests or timeouts if Prometheus is temporarily unavailable, anything like that. We (or rather the client library) just write and serve the metric data in the appropriate format and Prometheus handles the rest.
```

sarayourfriend commented:
> Is there a way (or use) to add monitoring for the size of the responses that we send?

Yeah totally, that would be a [histogram](https://prometheus.io/docs/concepts/metric_types/#histogram) metric that we could implement with relative ease.

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+## Purpose
+
+This RFC proposes we introduce [Prometheus](https://prometheus.io) and [Grafana](https://grafana.com/) to monitor our infrastructure. It also proposes that we use our existing Postgres instance for storing the gathered metrics.
```
```suggestion
This RFC proposes we introduce [Prometheus](https://prometheus.io) and [Grafana](https://grafana.com/) to monitor our infrastructure. It also proposes that we create a new RDS Postgres instance for Grafana.
```

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+#### Runbooks
+
+Runbooks are the expected steps for addressing a particular alarm. As I mentioned elsewhere, alarms should be actionable. If they're not actionable then they're just noise and should be demoted to a warning until they're more accurate (less false positives). The _actions_ you take for any given alarm are directed by an alarm's accomanying runbook.
```
```suggestion
Runbooks are the expected steps for addressing a particular alarm. As I mentioned elsewhere, alarms should be actionable. If they're not actionable then they're just noise and should be demoted to a warning until they're more accurate (less false positives). The _actions_ you take for any given alarm are directed by an alarm's accompanying runbook.
```

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+* Use ECS to deploy the following services in production:
+    * Prometheus
+    * Grafana
```
Note that Grafana is already deployed in our production infrastructure. I'm not sure if we'll replace or reuse the existing one. Probably whichever is less work (presumably reusing).

cc @zackkrida for final decision there.

zackkrida replied:
The current service is so simple, we should just be able to update the grafana version and redepoy. That's what I recommend. We can look at switching to ECS later.

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+## Overview of proposed changes
+
+* Use ECS to deploy the following services in production:
+    * Prometheus
+    * Grafana
```
```suggestion
* Use the existing Grafana instance in production
* Use ECS to deploy the following services in production:
    * Prometheus
```

sarayourfriend commented on rfcs/20220307-monitoring_across_the_stack.md:
```diff
@@ -0,0 +1,355 @@
+
+- [x] @zackkrida (Infrastructure sign off)
+- [x] @AetherUnbound
+- [ ] Approver TBD (API/Frontend person)
```
```suggestion
- [ ] @obulat 
```
